### ISSUE TITLE
Implement Predecessor Operation

### ISSUE THREAD

**Issue by Bonifatius94** [Bonifatius94/veb-in-c#15]
## /refine #15 — Implement Predecessor Operation

**Goal.** Replace the `assert(false)` stub in `vebtree_predecessor` with a working CLRS-style implementation that mirrors `vebtree_successor`, accounting for the one structural asymmetry: `tree->low` is not stored in any subtree, so it's a valid predecessor candidate when global lookup returns null.

**Affected code.**
- `include/vebtrees.h:262-268` — fix the pre-existing bit-0 conflation bug in `vebtree_bitwise_leaf_predecessor` **first**, before implementing the main function. The internal recursion into `pred(global, 0)` triggers this bug on any external query where `global_key == 0` (roughly: any `key < 64`), so the main function cannot be tested without the fix.
- `include/vebtrees.h:470-475` — replace the `assert(false)` stub with the real recursive implementation.
- `test/unit_tests.c:82-88` — register new predecessor test cases in `main`.
- `test/sorting_benchmark.c:13-28` — add a `sort_veb_pred` variant that sorts descending via `vebtree_get_max` + `vebtree_predecessor`, reverse-checked against the successor output.

**Approach.** Ordered as a strict red-green-refactor sequence:

1. **Leaf-predecessor fix first.** `vebtree_bitwise_leaf_predecessor` returns `max_bit_set(0)` when `pred_bits == 0` — which is `0xFF` on GCC/fallback (UB treated as `64 - 64 - 1`) and `0` on MSVC, so the existing `max_pred == 0 && (low & 1) == 0` sentinel only works on MSVC. Fix by checking `pred_bits == 0` up front, matching the pattern in `vebtree_bitwise_leaf_successor` at `vebtrees.h:254`.
2. **Empty tree + leaf delegation.** Make `vebtree_predecessor` handle the leaf case and the empty-internal case.
3. **Core recursive body.** Mirror `vebtree_successor` at `vebtrees.h:443`:
   - Fast path `if (tree->high != vebtree_null && key > tree->high) return tree->high;`.
   - Split `key` into `local_key` / `global_key` via `vebtree_local_address` / `vebtree_global_address`.
   - Local-hit: if `locals[global_key]` is non-empty and `local_key > vebtree_get_min(&locals[global_key])`, recurse into the local and prepend `global_key << lower_bits`.
   - Else recurse `global_pred = vebtree_predecessor(tree->global, global_key)`; if non-null, answer is `(global_pred << lower_bits) | vebtree_get_max(&locals[global_pred])`.
4. **Asymmetric tail.** When `global_pred == vebtree_null`, return `tree->low` if `key > tree->low`, else `vebtree_null`. This is the one part that isn't a mirror of `successor` — `low` is not stored in any subtree.
5. Declare all locals at the top of the function (C89 rule). Keep the existing `/* recursion anchor */` / `/* recursion case */` comment convention.

**Invariants to preserve** (from CLAUDE.md):
- `low` of an internal node is never in a subtree — this is precisely what drives the asymmetric tail step.
- `tree->high` equals the max key when non-empty; the predecessor path is read-only, must not mutate it.
- Empty-check semantics unchanged: internal node ⇔ `low == vebtree_null`; leaf ⇔ bitboard `low == 0`.
- No new dependencies; C89 only; no `//` comments; mixed decls/statements forbidden.

**Test plan.** One RGR cycle per step above, each with a test that fails before the step and passes after:

- **Cycle 1 — leaf fix.** `should_handle_bit_zero_in_bitwise_leaf_predecessor` — build a leaf tree (u=6), insert bit 5, assert `vebtree_predecessor(leaf, 1) == vebtree_null`. Currently returns `0xFF` (GCC). Also cover: insert bit 0, assert `predecessor(1) == 0` (stays correct). After fix, both pass.
- **Cycle 2 — empty + leaf delegation in public API.** `should_return_null_on_empty_tree_u4096` — empty u=12 tree, predecessor of any key is null. Also `should_delegate_to_leaf_in_small_tree` — u=6 tree with keys, predecessor behaves correctly.
- **Cycle 3 — core recursive path.** `should_find_predecessor_in_fully_alloc_tree_u4096` — insert `0..4095`, assert `vebtree_predecessor(tree, k) == k-1` for `k in 1..4095` and `predecessor(0) == vebtree_null`. This depends on the leaf fix being in place; without it, `pred(k)` for `k < 64` corrupts via `locals[0xFF]` out-of-bounds.
- **Cycle 4 — asymmetric tail.** `should_find_predecessor_crossing_low` — insert `{5, 200}` in u=12, assert `predecessor(50) == 5`. Forces the `global_pred == null` → `tree->low` fallback. Without the fallback, returns `null`.
- **Regression / coverage batch (non-RGR commit).** Passes without new code once Cycle 4 lands; bundled into one `test:` commit:
  - `should_find_predecessor_with_gaps_u4096` — insert odd keys only; predecessor of even is `k-1`, predecessor of odd `k > 1` is `k-2`.
  - `should_find_predecessor_on_singleton` — insert `{k}`, queries above return `k`, queries `<= k` return null.
  - `sort_veb_pred` in the sorting benchmark — starts at `vebtree_get_max`, walks predecessors, asserts reverse-order of `sort_veb_succ`.

**Risks / open questions.**
- The original plan tried to split the work into "avoid bit 0" (Phase 1) + "fix leaf" (Phase 2). That doesn't work: the leaf bug triggers on *internal* recursive calls into `pred(global, 0)`, which fire for any external query with `key < 64`, regardless of whether the answer itself is bit 0. Hence the re-ordering here — leaf fix goes first.
- `vebtree_predecessor(tree, vebtree_null)` asserts, matching `insert`/`delete`/`contains`.
- Duplicate-key / missing-key safety is issue #18. Predecessor on a tree corrupted by double-insert is out of scope — assume well-formed inputs, consistent with the rest of the API.

**Dependencies** (per ROADMAP.md):
- Must land first: none.
- Unblocks: reverse-sort validation of the sorting benchmark; no hard dependency from later roadmap items, but this is the first item in the suggested execution order so #18 and #5 follow.

**Comment by Bonifatius94** [comment]
## Implementation notes / deviations from the refined plan

Implemented on `main` via seven commits (`1c869ad..5f212b0` red/green/test). Three deviations worth flagging:

1. **Cycles 3 & 4 merged into one green commit** (`4ed42fd`). The dense 0..4095 test requires `pred(1) == 0`, which implicitly needs the low-fallback. Splitting the recursion body and the asymmetric tail into separate greens wasn't possible without contorting the RED test. The crossing-low test moved to the regression commit.
2. **No refactor commits.** Each green was already minimal and mirrored the established `vebtree_successor` structure. Nothing internal to extract without touching unrelated code.
3. **Added `build-debug.sh`** as scaffolding in the first RED commit. The default `build.sh` compiles with `-O3 -DNDEBUG`, which strips `assert()` — meaning every existing unit test has been silently running as a smoke test, not a correctness check. The new script builds Debug so assertions fire. Release path via `build.sh` is untouched.

**Follow-up worth tracking as its own issue:** the NDEBUG-in-release issue above. It's orthogonal to #15 but means none of the pre-existing `should_insert_into_fully_alloc_tree_u4096` / `should_delete_from_fully_alloc_tree_u4096` tests have ever actually failed on a regression — they've been checking nothing.

**Comment by Bonifatius94** [comment]
Implemented via commits 1c869ad..5f212b0 on main; test-asserts/benchmark cleanup in 7364f2e, e8c9f9b, 332df0c. Unit tests + SortingBenchmark green on Release (asserts preserved in UnitTests).
